### ISSUE TITLE
Check pkgdown grouping

### ISSUE THREAD

**Issue by maelle** [ropensci-review-tools/pkgcheck#100]
Via https://github.com/ropensci/dev_guide/issues/352#issuecomment-948483339

I.e. the existence of an entry "reference" in the pkgdown config. Maybe not a fail/pass but rather some sort of information, "the package has X reference pages organized in Y topic(s)" or so.

I'm still not sure it's worth an automatic check as I feel the usefulness of the reference index organization might better be assessed by an human.

**Comment by mpadge** [comment]
@maelle Those commits close this through implementing a simple check that there is some kind of grouping used in `man` entries, through collating all distinct `\concept{...}` entries, and ensuring there are at least 2 of those. This is contrary to what i said above, because is a check with a summary method only, and no print method. It also *requires* that people group functions by concept - [current wording in DevGuide](https://devguide.ropensci.org/building.html#roxygen2-use) is that

> We recommend using the `@family` tag ...

but that is then not a strict requirement. I'll re-open this issue to ask you whether this test should maybe be switched off removing the `summary` method on fail? Alternatively, just close issue if you're okay with this check.

The commits above add all the infrastructure to provide information on concept groups which can be used for any subsequent refinement, such as printing out groups for editors/reviewers to see, or whatever else may be desired down the line.

**Comment by mpadge** [comment]
Discussions on rOpenSci slack mostly indicated view that this should be *recommended* yet not *required.*. Given that, check will be switched off, yet infrastructure retained to enable any future checks to use the same info.

**Comment by maelle** [comment]
Thanks!

**Comment by mpadge** [comment]
@maelle I'm trying to clear out a few issues here. This check could potentially be replaced with `pkgdown::check_pkgdown()`, but that function is purely called for side-effects, and neither actually does anything functional nor returns any useful info. I could implement a work-around here, but that would be better done on `pkgdown` side. Thoughts?

**Comment by maelle** [comment]
`pkgdown::check_pkgdown()` won't tell you if you have 100 functions but no `reference:` configuration.

But anyway the reference grouping is in the editor template: https://devguide.ropensci.org/editortemplate.html

**Comment by mpadge** [comment]
Thanks, but what I mean is that `check_pkgdown()` could be revised to actually return useful information. It would be better to use that instead of hand-coded info extraction here. Otherwise I guess we could just tweak the local checks to use a threshold and issue maybe a failure if > 10 functions but no reference grouping?

**Comment by maelle** [comment]
Or we rely on the editor's template, to me that's fine.

**Comment by mpadge** [comment]
I'd rather implement a local check, because I do think we can and should do everything we can to make editors' work easier. This is definitely a case where we can totally automate those checks. So I'll leave it open to extend and refine the check a bit, and I'll also update the editors template to remove that item once it's implemented here. Thanks

**Comment by maelle** [comment]
One easy thing that'd be helpful to me is a direct link to the reference to click in the report 😂

**Comment by mpadge** [comment]
I suppose the report _could_ grab any "concept" groups from `man/` entries, and create a table from that? Not sure where it would go? Either

- In the report, with downside that an already very long and complex report would get even more so
- In some external file, uploaded like the network diagrams, with downside that we're then creating even more extraneous guff which represents more moving parts to a complicated system.

No easy solution, but I'll have a think about it.

**Comment by maelle** [comment]
Oh but isn't it fine if only the pkgdown reference page is organized, without using concepts in the manual?

To me just seeing the reference index is enough to judge whether some thought has been put into it, which is what the edit should check.

**Comment by mpadge** [comment]
The only option to organize a `pkgdown` ref page is by `@family` tags, and those are always and unavoidably rendered as `concepts` in `.Rd` files. So they're identically the same thing. I think what I should therefore do is simply activate the check like I suggested above, using a threshold that requires family/concept grouping if `num_rds > 10`. It's currently actually not active at all:
https://github.com/ropensci-review-tools/pkgcheck/blob/7c503f3caf60fdeea2f70ff5b753963d598d20b3/R/check-pkgdown.R#L12-L16

**Comment by maelle** [comment]
> The only option to organize a pkgdown ref page is by `@family` tags

No, many people organize the reference pages more manually with topic names, even if the method you state is very handy. https://pkgdown.r-lib.org/reference/build_reference.html#reference-index

**Comment by mpadge** [comment]
Oh I see. Sorry for my confusion. That makes it really tricky. Especially so in this case because `pkgdown` just passes everything through to [`config_abort()`](https://github.com/r-lib/pkgdown/blob/f69b62a7e74b58b42170284b6f3f56c674e7a3f8/R/config.R#L236-L246) which goes straight to `cli_abort()`. That mashes all input via `glue` so all useful information is simply destroyed. Grrr .... 😠 

Now I really need your opinion here 🙏 : Should i ...

- [ ] Grab and use internal `pkgdown` functions via `utils::getFromNamespace()`, mainly `as_pkgdown()` and `config_pluck_reference()`?
    - Upside: Stays aligned with `pkgdown`, and will follow any changes to topic identification mechanisms
    - Downside: Yet another dependency to an already dep-heavy package
- [ ] Rewrite local equivalents of the code logic of those bits in this package, and so avoid adding `pkgdown` as another dep here?
    - Upside: No extra dependency
    - Downside: `pkgdown` mechanisms could drift from here

I'd prefer the 2nd, but that would be more reliant on you to ensure no drift between the two.

**Comment by maelle** [comment]
Why not other lower-effort possibilities:

- Just check there's `reference:` in the pkgdown config
- Add the direct link to the reference page for the editor/reviewer to click

Also, I forget: does pkgcheck check that the pkgdown website can be built (info it could get from https://ropensci-staging.r-universe.dev/packages)? If it can be built, it means the reference entry in the pgkdown config does not have any missing topic which is good.

**Comment by mpadge** [comment]
> -  Just check there's `reference:` in the pkgdown config

But that's also not a requirement - I've got one pkg that doesn't even have or need a `_pkgdown.yaml` file, because it "just works". And if it did have one, I'd only do it to add aria labels to images, and still wouldn't need a `reference` section. That's the origin of my notion of using a `num_rd`-type threshold.

> - Add the direct link to the reference page for the editor/reviewer to click

That's a possibility, but also might not always exist. We've had a couple of submissions for which the main docs were books, while the only reference indices were on the CRAN manuals. And `pkgdown` is only a recommendation, not a strict requirement, which means a reference page is strictly speaking also only a recommendation. That's why my thoughts all tend towards `if (pkgdown)` -> then dump information that are more usable version of `check_pkgdown()` would actually deliver.

> Also, I forget: does pkgcheck check that the pkgdown website can be built (info it could get from [ropensci-staging.r-universe.dev/packages](https://ropensci-staging.r-universe.dev/packages))? If it can be built, it means the reference entry in the pgkdown config does not have any missing topic which is good.

That's a great idea!!! But also, very unfortunately, can't generally be done because we won't ever have a requirement _on submission_ that packages even exist on any universe. And short of that, one of the most viable paths remains `if (pkgdown)` -> `check_pkgdown()`.

Any further insights?

**Comment by maelle** [comment]
In the next dev guide release, pkgdown websites will actually be required at submission: https://devdevguide.netlify.app/booknews.html 

Your threshold idea is fine, but couldn't it be if more than N manual pages, there must be a pkgdown config with a reference entry?

**Comment by maelle** [comment]
actually just website, sorry. so maybe check the docs website is up and running, and if pkgdown, use the threshold + reference entry existence?

**Comment by maelle** [comment]
Further thought: since we build pkgdown websites after approval, shouldn't we require a pkgdown website can be built?

**Comment by mpadge** [comment]
> In the next dev guide release, pkgdown websites will actually be required at submission: [devdevguide.netlify.app/booknews.html](https://devdevguide.netlify.app/booknews.html)

I got temporarily excited about that and thought, wow, really, I hadn't seen that. Looked at booknews, saw link to [1004](https://github.com/ropensci/dev_guide/pull/1004), went there, and saw it was from me 😆 . Anyway, that just requires "a website", and we still just
> recommend using [`pkgdown`](https://github.com/r-lib/pkgdown) to create your website.

So still makes this tricky. And we also can't check that a website can be built because `check_pkgdown()` errors when no `_pkgdown.yaml` can be found, even though the site will build fine anyway.

---

## Recommendation

To return to the aim here: To be able to automatically satisfy this currently manual requirement for editors:

>  Is the reference index page clear (grouped by topic if necessary)?

And i think "clear" should always be anything at all if there are few functions. If there are "several" or "many" functions, then clear should be some kind of topic grouping. So I suggest that takes us to:

```
if (num_rds <  threshold) {
  # all is okay
} else if (pkgdown) {
  # some kind of `_pkgdown.yaml`-specific checking, like `check_pkgdown()`
}
```

**Comment by maelle** [comment]
>  # some kind of `_pkgdown.yaml`-specific checking, like `check_pkgdown()`

Yes, 

- `check_pkgdown()`
- having a `reference:` field
- counting on the editor/reviewer to look at the grouping?

**Comment by mpadge** [comment]
@maelle All down, with these two distinct diagnoses:
https://github.com/ropensci-review-tools/pkgcheck/blob/c4743dfdd1c8516b0fdebc861818ebba9d84615f/R/check-pkgdown.R#L17-L21
And those are only triggered if number of `.Rd` files is >= 10, otherwise no `_pkgdown.yml` expected at all, so no grouping.

**Comment by maelle** [comment]
Thank you!!
